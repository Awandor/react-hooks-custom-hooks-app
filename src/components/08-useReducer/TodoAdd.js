import { useForm } from "../../hooks/useForm";

export const TodoAdd = ( { handleAddTodo } ) => {

    const [ { description }, handleInputChange, resetForm ] = useForm( {
        description: '' // tiene que ser el nombre del input
    } );

    const handleSubmit = ( e ) => {

        e.preventDefault(); // evitar el refresco de la p√°gina

        if( description.trim().length <= 1 ){

            return;
        
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddTodo( newTodo );

        resetForm();

    };

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input type="text" name="description" placeholder="Algo" autoComplete="off" value={description} onChange={handleInputChange} />

                <button className="btn btn-success mt-2 btn-block" type="submit">Agregar</button>
            </form>
        </>
    );

};
