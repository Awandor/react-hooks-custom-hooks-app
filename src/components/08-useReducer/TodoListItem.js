export const TodoListItem = ( { index, todo, handleToggle, handleDelete } ) => {

    console.log( '  Me volv√≠ a generar :(  ' );

    return (
        <li key={todo.id} className="list-group-item">
            <span className={ `${todo.done && 'complete'}` } onClick={() => handleToggle( todo.id )}>{index + 1}. {todo.desc}</span>
            <button className="btn button-small btn-danger" onClick={() => handleDelete( todo.id )}>Borrar</button>
        </li>
    );

};
