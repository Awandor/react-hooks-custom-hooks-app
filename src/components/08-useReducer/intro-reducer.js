// Creamos un estado inicial

const initialState = [ {
  id: 1,
  todo: 'Comprar pan',
  done: false
} ];

// Creamos un reducer, que recibe el state y la acción que va a modificar el state y por defecto siempre vamos a retornar el state
// al argumento state le damos por defecto el valor de initialState

const todoReducer = ( state = initialState, action ) => {

  // La primera vez que se ejecuta no hay action por lo que debemos ponerla como opcional en el if

  if ( action?.type === 'agregar' ) {

    // Esta acción no puede ser asíncrona ni depender de cosas externas tiene que resolverse con lo que se tiene internamente

    return [ ...state, action.payload ];

  }

  return state;

}

// Esto es un reducer porque retorna algo

let todos = todoReducer();

// Supongamos que quiero añadir un todo al arreglo de todos

// No podemos usar push porque modificaría el arreglo y por tanto modificaría el state

// En general no se usa push cuando trabajamos con React

const newTodo = {
  id: 2,
  todo: 'lavar el coche',
  done: false
};

// Vamos a enviar el newTodo como acción a todoReducer, no lo hacemos diréctamente sino a través de un objeto

const agregarTodoAction = {
  type: 'agregar', // siempre tiene que tener la propiedad type
  // newTodo
  payload: newTodo // se puede poner cualquier nombre a la propiedad pero es habitual payload o el mismo objeto
};

todos = todoReducer( todos, agregarTodoAction );

console.log( todos );
