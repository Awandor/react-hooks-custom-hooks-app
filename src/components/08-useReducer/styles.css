.list-group-item {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

span {
  cursor: pointer;
}

.complete {
  text-decoration: line-through;
}