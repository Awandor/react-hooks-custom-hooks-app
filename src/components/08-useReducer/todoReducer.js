export const todoReducer = ( state = [], action ) => {

  switch ( action.type ) {
    case 'add':
      return [ ...state, action.payload ];

      // break; // No hace falta break porque retornamos

    case 'delete':

      // console.log( action.payload );

      // filter retorna un nuevo arreglo

      return state.filter( todo => todo.id !== action.payload );

    case 'toggle':

      /* return state.map( todo => {

        if ( todo.id === action.payload ) {

          return {...todo, done: !todo.done };

        } else {

          return todo; // Si no retornamos nada el map va a retornar undefined

        }

      } ); */

      return state.map( todo => ( todo.id === action.payload ) ? {...todo, done: !todo.done } : todo );

    default:
      return state; // si no hay acci√≥n retorna el estado
  }

};
