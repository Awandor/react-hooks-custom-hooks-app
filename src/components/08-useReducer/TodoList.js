import { TodoListItem } from "./TodoListItem";

export const TodoList = ( { todos, handleToggle, handleDelete } ) => {

    // console.log( 'TodoList', todos );
    // console.log( 'handleToggle', handleToggle );

    return (
        <ul className="list-group list-group-flush">
            {

                /* todos.map( ( todo, i ) => (
                    <li key={todo.id} className="list-group-item">
                        <span className={ `${todo.done && 'complete'}` } onClick={() => handleToggle( todo.id )}>{i + 1}. {todo.desc}</span>
                        <button className="btn button-small btn-danger" onClick={() => handleDelete( todo.id )}>Borrar</button>
                    </li>
                ) ) */

                todos.map( ( todo, i ) => <TodoListItem key={todo.id} index={i} todo={todo} handleToggle={handleToggle} handleDelete={handleDelete} /> )
            }
        </ul>
    );

};
