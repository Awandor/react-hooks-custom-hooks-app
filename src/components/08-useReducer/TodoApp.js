import { /* useCallback, */ useEffect, useReducer } from 'react';
import './styles.css';
import { todoReducer } from './todoReducer';

// import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

/* const initialState = [ {
    id: new Date().getTime(),
    desc: 'Aprender React',
    done: false
} ]; */

const init = () =>{

    /* return [ {
        id: new Date().getTime(),
        desc: 'Aprender React',
        done: false
    } ]; */

    return JSON.parse( localStorage.getItem( 'todos' ) ) || []; // para manejar el null

};

export const TodoApp = () => {

    // 1. Creamos el useReducer
    // reducer el la función reducer vista anteriormente
    // initialstate es el estado inicial que podemos mandar
    // init es una función para inicializar el state en caso de que ese state sea procesado o haga varias acciones, vamos a hacerlo con y sin init
    // dispatch ejecuta acciones hacia el reducer, si hacemos cambios en el state React reacciona redibujando en pantalla el cambio

    // const [state, dispatch] = useReducer(reducer, initialState, init);

    // 2. Empezamos de manera más sencilla y creamos initialState fuera del componente

    // 3. El reducer lo creamos en un archivo independiente pues puede crecer mucho

    // 4. En la desestructuración ponemos nombre todos

    // const [ todos, dispatch ] = useReducer( todoReducer, initialState );
    const [ todos, dispatch ] = useReducer( todoReducer, [], init );

    // console.log( 'todos', todos );

    // custmo Hook useForm, handleInputChange lo colocamos en el input

    /* const [ formValues, handleInputChange ] = useForm( {
        description: ''
    } ); */

    // console.log( 'formValues', formValues );

    // Como formValues es un objeto con la propiedad description podemos aplicar desestructuración

    /* const [ { description }, handleInputChange, resetForm ] = useForm( {
        description: '' // tiene que ser el nombre del input
    } ); */

    // Nos lo llevamos a TodoApp.js

    useEffect( ()=>{

        // localStorage sólo guarda strings

        localStorage.setItem( 'todos', JSON.stringify( todos ) );

    }, [ todos ] );


    // console.log( 'description', description ); // Se lo aplicamos al value del input

    /* const handleSubmit = ( e ) => {

        e.preventDefault(); // evitar el refresco de la página

        // Validamos que haya contenido

        if( description.trim().length <= 1 ){

            return;
        
        }

        console.log( description );

        // Objeto nueva tarea

        const newTodo = {
            id: new Date().getTime(),

            // desc: 'Nueva tarea',
            desc: description, // podemos establecerlo directamente porque cada vez que se hace un cambio React vuelve a ejecutar TodoApp
            done: false
        }

        // Acción que se envía al reducer

        const action = {
            type: 'add',
            payload: newTodo
        };

        dispatch( action );

        resetForm();

    } */

    const handleDelete = ( todoId ) => {

        console.log( 'handleDelete', todoId );

        const action = {
            type: 'delete',
            payload: todoId
        };

        dispatch( action );

    };

    const handleToggle = ( todoId ) => {

        console.log( 'handleToggle', todoId );

        const action = {
            type: 'toggle',
            payload: todoId
        };

        dispatch( action );

    };

    // Parece que no hace falta usar useCallback

    /* const handleDelete = useCallback(
        ( todoId ) => {

            dispatch( { type: 'delete', payload: todoId } );

        },
        [ ]
    );

    const handleToggle = useCallback(
        ( todoId ) => {

            dispatch( { type: 'toggle', payload: todoId } );

        },
        [ ]
    ); */

    const handleAddTodo = ( newTodo )=>{

        dispatch( { type:'add', payload: newTodo } );

    };

    return (
        <div className="container">
            <h1>Todo App</h1>
            <hr />

            <h2 className="mb-3">Nº de tareas <small>{todos.length}</small></h2>

            <div className="row">
                <div className="col-7">

                    {/* <ul className="list-group list-group-flush">
                        {
                            todos.map( ( todo, i ) => (
                                <li key={todo.id} className="list-group-item">
                                    <span className={ `${todo.done && 'complete'}` } onClick={() => handleToggle( todo.id )}>{i + 1}. {todo.desc}</span>
                                    <button className="btn button-small btn-danger" onClick={() => handleDelete( todo.id )}>Borrar</button>
                                </li>
                            ) )
                        }
                    </ul> */}

                    <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete} />

                </div>
                <div className="col-5 text-center">
                    {/* <h4>Agregar TODO</h4>
                    <hr />
                    <form onSubmit={handleSubmit}>
                        <input type="text" name="description" placeholder="Algo" autoComplete="off" value={description} onChange={handleInputChange} />

                        <button className="btn btn-success mt-2 btn-block" type="submit">Agregar</button>
                    </form> */}

                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    );

};
