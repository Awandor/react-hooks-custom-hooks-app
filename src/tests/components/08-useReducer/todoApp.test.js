import { act } from 'react-dom/test-utils';
import { mount, shallow } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';

describe( 'Prueba en <TodoApp />', () => {

    const wrapper = shallow( <TodoApp /> );

    Storage.prototype.setItem = jest.fn();

    test( 'debe mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();

    } );

    test( 'debe agregar un TODO', () => {

        // Esto ya ha sido probado pero lo necesitamos para evaluar otras cosas como el local storage

        // Vamos a usar mount en vex de shallow, shallow es básico y se centra en el componente, mount
        // va más allá y tiene en cuenta los componente hijos y el entorno

        // Aunque funciona igual con shallow

        const wrapper2 = mount( <TodoApp /> );

        const addTodo = wrapper2.find( 'TodoAdd' ).prop( 'handleAddTodo' );

        // Tenemos que envolver addTodo con act pues estamos cambiando el state

        act( () => {

            addTodo( demoTodos[0] );

            addTodo( demoTodos[1] );

        } );

        expect( wrapper2.find( 'h2' ).text().trim() ).toBe( 'Nº de tareas 2' );

        expect( localStorage.setItem ).toHaveBeenCalledTimes( 2 ); // Parece que es llamado 1 vez al iniciarse el componente

    } );

    test( 'debe eliminar un TODO', () => {

        // Usamos el wrapper original con shallow y funciona igualmente

        const addTodo = wrapper.find( 'TodoAdd' ).prop( 'handleAddTodo' );

        addTodo( demoTodos[0] );

        const deleteTodo = wrapper.find( 'TodoList' ).prop( 'handleDelete' );

        deleteTodo( demoTodos[0].id );

        expect( wrapper.find( 'h2' ).text().trim() ).toBe( 'Nº de tareas 0' );

    } );



} );
