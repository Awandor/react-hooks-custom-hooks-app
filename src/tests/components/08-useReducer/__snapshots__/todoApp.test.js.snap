// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <TodoApp /> debe mostrarse correctamente 1`] = `
<div
  className="container"
>
  <h1>
    Todo App
  </h1>
  <hr />
  <h2
    className="mb-3"
  >
    NÂº de tareas 
    <small>
      0
    </small>
  </h2>
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5 text-center"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
